go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -115,25 +124,130 @@ func (c *Controller) UserDefinedNetworkReconciler(key string) error {
+	}
+
+	if !reflect.DeepEqual(nad.Spec, desiredNAD.Spec) ||
+		!reflect.DeepEqual(nad.ObjectMeta.OwnerReferences, desiredNAD.ObjectMeta.OwnerReferences) {
+		nad.Spec.Config = desiredNAD.Spec.Config
+		nad.ObjectMeta.OwnerReferences = desiredNAD.ObjectMeta.OwnerReferences
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nad.Namespace).Update(context.Background(), nad, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to update NetworkAttachmetDefinition: %w", err)
+		}
+		klog.Infof("Updated NetworkAttachmetDefinition [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	return nad, nil
+}
+
+func ownedByUserDefinedNetwork(nad *netv1.NetworkAttachmentDefinition, gvk schema.GroupVersionKind, uid types.UID) bool {
+	for _, ownerRef := range nad.OwnerReferences {
+		if ownerRef.UID == uid && ownerRef.Kind == gvk.Kind && ownerRef.APIVersion == gvk.GroupVersion().String() {
+			return true
+		}
+	}
+	return false
 }
 
 func (c *Controller) UpdateUserDefinedNetworkStatus(udn *userdefinednetworkv1.UserDefinedNetwork, nad *netv1.NetworkAttachmentDefinition, syncError error) error {
-	//TODO: implement UDN status update logic
-	return errors.New("implement me")
+	if udn == nil {
+		return nil
+	}
+
+	networkReadyCondition := newNetworkReadyCondition(nad, syncError)
+
+	var updatedConditions []metav1.Condition
+	if len(udn.Status.Conditions) == 0 {
+		updatedConditions = append(updatedConditions, *networkReadyCondition)
+	} else {
+		cp := slices.Clone(udn.Status.Conditions)
+		updatedConditions = updateCondition(cp, networkReadyCondition)
+	}
+
+	if !reflect.DeepEqual(udn.Status.Conditions, updatedConditions) {
+		var err error
+		udnApplyConf := udnapplyconfkv1.UserDefinedNetwork(udn.Name, udn.Namespace).
+			WithStatus(udnapplyconfkv1.UserDefinedNetworkStatus().
+				WithConditions(updatedConditions...))
+		opts := metav1.ApplyOptions{FieldManager: "user-defined-network-controller"}
+		udn, err = c.udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).ApplyStatus(context.Background(), udnApplyConf, opts)
+		if err != nil {
+			return fmt.Errorf("failed to update UserDefinedNetwork status: %w", err)
+		}
+		klog.Infof("Updated status UserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
+	}
+
+	return nil
+}
+
+func newNetworkReadyCondition(nad *netv1.NetworkAttachmentDefinition, syncError error) *metav1.Condition {
+	now := metav1.Now()
+	networkReadyCondition := &metav1.Condition{
+		Type:               "NetworkReady",
+		Status:             metav1.ConditionTrue,
+		Reason:             "NetworkAttachmentDefinitionReady",
+		Message:            "NetworkAttachmentDefinition has been created",
+		LastTransitionTime: now,
+	}
+
+	if nad == nil {
+		networkReadyCondition.Status = metav1.ConditionFalse
+		networkReadyCondition.Reason = "NetworkAttachmentDefinitionNotExist"
+		networkReadyCondition.Message = "NetworkAttachmentDefinition is not exist"
```

Comment:
I am not sure if this reason is useful, I would like to know why it doesn't exist, which is probably captured in the syncError, do you think we need this reason at all?

Reply:
now I think we don't need nadCopy to be returned by `syncUserDefinedNetwork`, as the only field we check in the status is deletionTimestamp, and it is set before `syncUserDefinedNetwork`, wdyt?

Reply:
In a scenario the UDN NAD not exist (copyNad=nil), `syncUserDefinedNetwork` will create the NAD and return the create object (copyNad points to the actual object).
When `syncUserDefinedNetwork` is changed to not return the NAD:
`nadCopy=nil` is passed to `syncUserDefinedNetwork` as an arg.
`syncUserDefinedNetwork` create the NAD and assign the created object to the given arg (nadCopy).
When `syncUserDefinedNetwork` returns, the assigned object is discarded and  `nadCopy` remain `nil`, although it was assigned with an actual object.

From what I understand, root case is due to `nil` value for for a object pointer argument.
When it assigned by the called function, the assigned object get discarded when the function returns because there is no actual memory allocated that persist the pointer value (because nil is given).

One way to fix this issue was changing `syncUserDefinedNetwork` signature to have NAD object as returned value.

Reply:
right, what I am saying is that we probably don't need the new non-nil nad object for `newNetworkReadyCondition` anymore, as it only cares about `nad.DeletionTimestamp` and it can't be non-zero if the original `nadCopy` was nil.
So in case we have created the nad, it is ok to pass nil to the `newNetworkReadyCondition`, as the condition will still be correct?